### PR TITLE
Expose Quality of Service settings (Priority, Congestion Control, Express) the Sample was published with

### LINKED ISSUE #694: Priority in samples
### Describe the feature

For mapping UAttributes onto Zenoh. It should be added on all languages

### PR DESCRIPTION
**Pull request by DenisBiryukov91**
Expose Quality of Service settings (Priority, Congestion Control, Express flag) the sample was sent with (closes #694)

### CODE REVIEW

File: zenoh/tests/qos.rs
Comment (Mallets) [inline zenoh/tests/qos.rs:31]: It's better to wait with a timeout with e.g. a [ztimeout!()](https://github.com/eclipse-zenoh/zenoh/blob/84dc04acdcf794e71667e0e76e521f41b921eef8/io/zenoh-transport/tests/unicast_transport.rs#L228) macro.

File: zenoh/tests/qos.rs
Comment (Mallets) [inline zenoh/tests/qos.rs:6]: I believe we should only use imports from `zenoh` (including `zenoh::prelude`) in the tests.
The only exception should be `zenoh_core::zasync_executor_init`.

File: zenoh/tests/qos.rs
Comment (Mallets) [inline zenoh/tests/qos.rs:42]: At API level there shouldn't be any `.into()` necessary for comparison. That means that we are exposing two different types and it shouldn't be the case.

File: zenoh/tests/qos.rs
Comment (Mallets) [inline zenoh/tests/qos.rs:33]: At API level there shouldn't be any .into() necessary for comparison. That means that we are exposing two different types and it shouldn't be the case.

**Review by Mallets — CHANGES_REQUESTED** [review]

File: zenoh/tests/qos.rs
Comment (Mallets) [inline zenoh/tests/qos.rs:22]: I find this very weird. 
`Priority` and `CongestionControl` should be set before calling `.res()` and not after.
I wonder if we should really allow that...

Comment (DenisBiryukov91) [reply zenoh/tests/qos.rs:39]: fixed

File: zenoh/tests/qos.rs
Comment (DenisBiryukov91) [reply zenoh/tests/qos.rs]: fixed

File: zenoh/tests/qos.rs
Comment (DenisBiryukov91) [reply zenoh/tests/qos.rs]: fixed

File: zenoh/tests/qos.rs
Comment (DenisBiryukov91) [reply zenoh/tests/qos.rs]: fixed

File: zenoh/tests/qos.rs
Comment (DenisBiryukov91) [reply zenoh/tests/qos.rs:1]: fixed

File: commons/zenoh-protocol/src/network/mod.rs
Comment (DenisBiryukov91) [reply commons/zenoh-protocol/src/network/mod.rs]: fixed

File: commons/zenoh-protocol/src/core/mod.rs
Comment (DenisBiryukov91) [reply commons/zenoh-protocol/src/core/mod.rs]: fixed

File: zenoh/src/publication.rs
Comment (Mallets) [inline zenoh/src/publication.rs:905]: This is not safe. `zenoh_protocol::core::Priority::Control` does not exist in `Priority`, so the mapping will not be possible in that case. It should be a `TryFrom`.

Comment (DenisBiryukov91) [reply zenoh/src/publication.rs]: fixed

File: zenoh/src/sample.rs
Comment (Mallets) [inline zenoh/src/sample.rs:532]: The only faillable conversion is `Priority`.
With current implementation the whole `QoS` is falling back to default while only the `Priority` field should be resetted to default. `congestion_control` and `express` fields shouldn't be changed.

Comment (DenisBiryukov91) [reply zenoh/src/sample.rs:532]: fixed

File: zenoh/src/sample.rs
Comment (Mallets) [inline zenoh/src/sample.rs:536]: I would put logging level to `trace` in this case. `error` may pollute too much.

Comment (DenisBiryukov91) [reply zenoh/src/sample.rs:536]: Given that we can fail priority conversion only if something really wrong and unlikely happened, would not it make more sense to keep it as error ?

Comment (DenisBiryukov91) [reply zenoh/src/sample.rs]: fixed

**Comment by Mallets** [comment]
It LGTM. However, given the recent discussion on `Sample` and `SourceInfo` I'd like to have the feedback also from @p-avital and @milyin .

**Review by Mallets — APPROVED** [review]

**Review by milyin — APPROVED** [review]

**Comment by Mallets** [comment]
Holding on in merging this PR since it will break the memory layout for zenoh-c and zenoh-cpp at least.
Sister PRs in zenoh-c and zenoh-cpp should be done pointing to this PR.

File: zenoh/src/sample.rs
Comment (p-avital) [inline zenoh/src/sample.rs:333]: I'm not a huge fan of how `Sample` keeps on getting more public fields. Maybe we can use the API break to make it a bit more private?

File: zenoh/src/sample.rs
Comment (p-avital) [inline zenoh/src/sample.rs:524]: Using accessors, this could be a bitfield and occupy a single byte instead of 3

**Review by p-avital — COMMENTED** [review]
I think it might be time (and our last chance) to move `Sample` to accessors.

**Review by p-avital — CHANGES_REQUESTED** [review]
@Mallets @OlivierHecart can you consider my proposal above of moving `Sample` to an accessor pattern?

File: zenoh/src/sample.rs
Comment (DenisBiryukov91) [reply zenoh/src/sample.rs:524]: What is the benefit of reducing 3  bytes to one on the system with word size >= 4 bytes ? As a separate struct it will still tend to occupy 4 bytes at least, no ?

Comment (p-avital) [reply zenoh/src/sample.rs:524]: It's all about breakpoints and, to be fair, I think this layout just barely fits in `Sample`'s current padding, but that also means the next field is guaranteed to make `Sample` fatter, which we've shown in previous layout improvements to be rather performances sensitive for us.

API-wise, it's tempting to have only public fields, but that turns any future change to the struct's fields an API break. Builder patterns are just as convenient and give us much more flexibility in the future, both for layout optimization and for field additions.

**Comment by Mallets** [comment]
> @Mallets @OlivierHecart can you consider my proposal above of moving `Sample` to an accessor pattern?

I think the proposal is in principle ok. The question is: should this PR be blocked by the proposal or should we merge it while we rework `Sample` entirely? I'm more in favour to get it in alongside the sister PRs so as to unblock some work being done in https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2.

**Comment by p-avital** [comment]
I'm fine with either, but I'd like at least `Qos` to come out with the accessor+builder API from the get-go.

Keep in mind that if we unblock them with a first branch, they'll have to deal with the breaking change immediately after.

**Comment by Mallets** [comment]
> I'm fine with either, but I'd like at least `Qos` to come out with the accessor+builder API from the get-go.
> 
> Keep in mind that if we unblock them with a first branch, they'll have to deal with the breaking change immediately after.

I'm fine with your proposal for Rust.

Similar PRs are available in:
- C https://github.com/eclipse-zenoh/zenoh-c/pull/244 https://github.com/eclipse-zenoh/zenoh-pico/pull/348
- C++ https://github.com/eclipse-zenoh/zenoh-cpp/pull/99
Let's be consistent as well with other bindings. Should we have some kind of accessors also in C and C++?

**Pull request merged**
